api/org/commcare/api/persistence/SqlIndexedStorageUtility.java

```diff
@@ -0,0 +1,418 @@
+            preparedStatement = c.prepareStatement(sqlQuery);
+            ResultSet rs = preparedStatement.executeQuery();
+            return rs.getInt(1);
+        } catch (Exception e) {
+            System.out.println("SqlIndexedStorageUtility readBytes exception: " + e);
+        } finally {
+            try {
+                if (c != null) {
+                    c.close();
+                }
+                if (preparedStatement != null) {
+                    preparedStatement.close();
+                }
+            } catch (SQLException e) {
+                e.printStackTrace();
+            }
+        }
+        return -1;
+    }
+
+    /* (non-Javadoc)
+     * @see org.javarosa.core.services.storage.IStorageUtility#getRecordSize(int)
+     */
+    public int getRecordSize(int id) {
+        //serialize and test blah blah.
+        return 0;
+    }
+
+    @Override
+    public SqlStorageIterator<T> iterate() {
+        Connection connection;
+        ResultSet resultSet = null;
+        PreparedStatement preparedStatement = null;
+        try {
+            connection = this.getConnection();
+            String sqlQuery = "SELECT " + org.commcare.modern.database.DatabaseHelper.ID_COL + " , " +
+                    org.commcare.modern.database.DatabaseHelper.DATA_COL + " FROM " + this.tableName + ";";
+            preparedStatement = connection.prepareStatement(sqlQuery);
+            resultSet = preparedStatement.executeQuery();
+        } catch (SQLException e) {
+            e.printStackTrace();
+        }
+        return new SqlStorageIterator<T>(preparedStatement, resultSet, this.getNumRecords(), this);
+    }
+
+    /* (non-Javadoc)
+     * @see org.javarosa.core.services.storage.IStorageUtility#getTotalSize()
+     */
+    public int getTotalSize() {
+        //serialize and test blah blah.
+        return 0;
+    }
+
+    /* (non-Javadoc)
+     * @see org.javarosa.core.services.storage.IStorageUtility#isEmpty()
+     */
+    public boolean isEmpty() {
+        return this.getNumRecords() > 0;
+    }
+
+    @Override
+    public byte[] readBytes(int id) {
+        PreparedStatement preparedStatement = null;
+        Connection c = null;
+        try {
+            c = getConnection();
+            preparedStatement = SqlHelper.prepareIdSelectStatement(c, this.tableName, id);
+            if (preparedStatement == null) {
+                return null;
+            }
+            ResultSet rs = preparedStatement.executeQuery();
+            return rs.getBytes(org.commcare.modern.database.DatabaseHelper.DATA_COL);
+        } catch (Exception e) {
+            System.out.println("SqlIndexedStorageUtility readBytes exception: " + e);
+        } finally{
+            try {
+                if (c != null) {
+                    c.close();
+                }
+                if (preparedStatement != null) {
+                    preparedStatement.close();
+                }
```

Comment:
The order on these closes are backwards. You should close the statement before the connection